kotlin/workflow-ui-android/src/main/java/com/squareup/workflow/ui/ModalViewContainer.kt

```diff
@@ -40,40 +39,59 @@ internal class ModalViewContainer
   private val modalDecorator: (View) -> View = { it }
 ) : ModalContainer<Any>(context, attributeSet) {
 
-  @ExperimentalWorkflowUi
-  override fun Any.matches(nextModal: Any) = this::class.java == nextModal::class.java
-
-  @ExperimentalWorkflowUi
-  override fun showDialog(
-    modalScreen: Any,
-    screens: Observable<out Any>,
+  override fun buildDialog(
+    initialValue: Any,
     viewRegistry: ViewRegistry
-  ): Dialog {
-    val binding = viewRegistry.getBinding<Any>(modalScreen::class.jvmName)
-    val view = binding.buildView(screens, viewRegistry, this)
+  ): DialogRef<Any> {
+    val view = viewRegistry.buildView(initialValue, this)
 
-    return Dialog(context, dialogThemeResId).apply {
-      setCancelable(false)
-      setContentView(modalDecorator(view))
-      window!!.setLayout(WRAP_CONTENT, WRAP_CONTENT)
+    return Dialog(context, dialogThemeResId)
+        .apply {
+          setCancelable(false)
+          setContentView(modalDecorator(view))
+          window!!.setLayout(WRAP_CONTENT, WRAP_CONTENT)
 
-      if (dialogThemeResId == 0) {
-        // If we don't set or clear the background drawable, the window cannot go full bleed.
-        window!!.setBackgroundDrawable(null)
-      }
-      show()
+          if (dialogThemeResId == 0) {
+            // If we don't set or clear the background drawable, the window cannot go full bleed.
+            window!!.setBackgroundDrawable(null)
+          }
+        }
+        .run {
+          DialogRef(initialValue, this, view)
+        }
+  }
+
+  override fun updateDialog(dialogRef: DialogRef<Any>): DialogRef<Any> {
```

Comment:
Just realized that I'm not actually using that return type. YAGNI, and I can imagine subclasses getting into a lot of confusion or trouble by making edits that I'm not set up to honor.